### ISSUE TITLE
configuration: formally consider `null` values as "option not provided"

### ISSUE THREAD

**Issue by hhromic** [vectordotdev/vector#12832]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Use Cases

We extensively use environment variables in all of our Vector pipelines to accomodate for different deployment cases. However, for optional configuration options, the user experience is not consistent. For example, consider the following YAML configuration:
```yaml
sources:
  source:
    type: stdin
sinks:
  sink:
    type: http
    inputs: ["source"]
    uri: ${HTTP_URI:-http://localhost:8080}
    encoding:
      codec: ndjson
    batch:
      max_bytes: ${HTTP_BATCH_MAX_BYTES:-null}
      max_events: ${HTTP_BATCH_MAX_EVENTS:-null}
      timeout_secs: ${HTTP_BATCH_TIMEOUT_SECS:-null}
    buffer:
      type: memory
      max_events: ${HTTP_BUFFER_MAX_EVENTS:-500}
    request:
      concurrency: ${HTTP_REQUEST_CONCURRENCY:-adaptive}
    tls:
      enabled: ${HTTP_TLS_ENABLED:-false}
      ca_file: ${HTTP_TLS_CA_FILE:-null}
      crt_file: ${HTTP_TLS_CRT_FILE:-null}
      key_file: ${HTTP_TLS_KEY_FILE:-null}
    proxy:
      enabled: ${HTTP_PROXY_ENABLED:-false}
      http: ${HTTP_PROXY_HTTP:-null}
      https: ${HTTP_PROXY_HTTPS:-null}
```

The above configuration allows us to easily and cleanly configure most of the sink operational parameters without having to modify the pipeline configuration itself. Neat!

However, as can be seen, we are unofficially _abusing_ the `null` value to indicate "use the Vector default here". This works for some options but not for everything. The `buffer.max_events` and `request.concurrency` options are good examples of this. In that case, we simply hard-code the Vector default in the variable expansion default.

Hard-coding Vector defaults like that is not always convenient. For example, the `batch.max_events` default is `usize::max_value()` from what [I could find](https://github.com/vectordotdev/vector/blob/623b18263ff4281c6c42cb9635a5403eb3991747/src/sinks/util/batch.rs#L246-L254) in the source code. This default looks very ugly as a variable expansion default: `18446744073709551615`. Luckily Vector accepts `null` as "use the default" in that case :)

Unfortunately, some other options do not like `null` for this purpose. An example is `proxy.no_proxy` (which we are not using in the above example configuration).

Moreover, since the release of Vector `0.21.x`, some options stopped accepting `null` as values. Notably, this is the case for the `tls.*` options. This breaks our deployments where we do not need TLS options configured. If we default to a dummy value such as `/path/to/file` it does not work because `tls.enabled=false` still tries to read any configured `*_file` option. We also tried defaulting to `/dev/null`, but still does not work because it still tries to read `/dev/null` and finds an invalid X509 certificate (which is quite expected). For this reason, we had to halt migrating to Vector `0.21.x` for the moment.

### Attempted Solutions

As described in the Use-case section, our best attempted solution to deal with optional configuration options is to abuse the `null` value as variable expansion default. However this does not work consistently and since Vector `0.21.x`, this does not work anymore in some options.

### Proposal

We think that a good solution would be to formally accept `null` values as a "use the Vector default value" for all single-unit configuration options, i.e. options that take booleans, integers, floats, strings, etc.

In other words, we propose that any `null` option value is functionally equivalent to not providing such option. We believe that something like this would be a nice UX improvement for Vector.

**Comment by spencergilbert** [comment]
I need to read this some more and give some more thoughtful feedback, but just wanted to note that we need to consider how this would work across TOML/JSON/YAML as I don't think TOML in particular doesn't support `null`.

**Comment by hhromic** [comment]
Thanks for taking the suggestion in consideration.
I also thought on the case of TOML which doesn't support `null` indeed as value. JSON and YAML both support `null` as values. My conclusion was that, for TOML configurations, it simply wouldn't be possible to set anything to `null` and would cause an error due to to the limitation of the syntax, not Vector itself. In that case, this feature would be only available for JSON/YAML configuration syntax. I'm not sure that would be a problem if documented.

Take your time to digest what I wrote and of course feel free to ask any questions for further discussion. It would be _really_ handy for us to have something like the proposal implemented in Vector.

**Comment by spencergilbert** [comment]
> Take your time to digest what I wrote and of course feel free to ask any questions for further discussion. It would be _really_ handy for us to have something like the proposal implemented in Vector.

Of course! cc'ing @jszwedko who mentioned also wanting to take time to think about it, he had some ideas - but has been busy this week working on the next release 😄

**Comment by jszwedko** [comment]
😄 

Yeah, I did want to reflect on this more.

Implementation-wise, I think it could look something like making every option that Vector has a default for an `Option<>` and doing the defaulting by using `unwrap_or(some default value)` when reading it rather than relying on `serde` defaulting. This should allow the use of `null` to say "I want the default value".

As Spencer points out, this wouldn't be usable in TOML which lacks `null`s but would work for JSON and YAML.

**Comment by jszwedko** [comment]
Also food for thought for @tobz since he's working on improving Vector's configuration.

**Comment by pront** [comment]
I opened #25981 to preserve explicit `null` values when loading JSON and YAML configuration. Previously, those formats were converted through TOML, which cannot represent `null`. The PR intentionally follows normal Serde semantics rather than making `null` universally equivalent to an omitted field:

- Omitted field → apply its configured default.
- `null` for `Option<T>` → `None`.
- Explicit value → deserialize that value.
- `null` for a non-optional field → return an error.

For example:

```yaml
timestamp_nanos_key: null
```

This deserializes to `None`, while `omitting timestamp_nanos_key` applies its default of `Some("@timestamp.nanos")`. This distinction is useful: treating null as omission would make it impossible to explicitly disable a defaulted optional value.

Therefore, #25981 supports native `null` semantics in YAML and JSON, but does not implement the proposal that every null means “use Vector’s default.” TOML remains unchanged because it has no `null` representation.

**Comment by hhromic** [comment]
Thanks for remembering this after so many years :)

To address our original use-case for this, which was to have a simple `option = ${ENV_OPTION}` setup and setting `ENV_OPTION=null` to mean "use whatever is the default for `option`", we resorted to use [`ytt`](https://github.com/carvel-dev/ytt) to preprocess variables and actually omit the YAML keys when the variables are empty or non existing. This creates the effect we were looking for at the cost of having to bundle a third-party YAML processing tool into our containers with Vector. Not a big deal actually and we got used to this with time. Moreover, `ytt` is quite powerful and allowed us to do very fancy and sophisticated things with Vector :)

Thanks again for working on this and feel free to close this issue with your PR !
I consider it "good enough" for this issue. 👍
